x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/Watcher.java

```diff
@@ -669,6 +671,30 @@ public String getFeatureName() {
         return "watcher";
     }
 
+    @Override
+    public void prepareForIndicesUpgrade(ClusterService clusterService, Client client, ActionListener<Boolean> listener) {
+        Client originClient = new OriginSettingClient(client, WATCHER_ORIGIN);
+        WatcherServiceRequest serviceRequest = new WatcherServiceRequest();
+        serviceRequest.stop();
+        originClient.execute(
+            WatcherServiceAction.INSTANCE,
+            serviceRequest,
+            ActionListener.wrap((response) -> { listener.onResponse(response.isAcknowledged()); }, listener::onFailure)
+        );
+    }
+
+    @Override
+    public void indicesUpgradeComplete(ClusterService clusterService, Client client, ActionListener<Boolean> listener) {
+        Client originClient = new OriginSettingClient(client, WATCHER_ORIGIN);
+        WatcherServiceRequest serviceRequest = new WatcherServiceRequest();
+        serviceRequest.start();
```

Comment:
What if watcher was _already_ stopped before system indices were upgraded?  Then upgrading system indices has the unforeseen side effect of enabling watcher.

ML will basically be doing the same as watcher with this API, except enabling and disabling ML upgrade mode instead of stopping and starting watcher.  So ML will have the same problem here too - if the user had already enabled ML upgrade mode for some other reason prior to upgrading system indices then upgrading system indices will have the side effect of disabling ML upgrade mode and allowing ML persistent tasks to start up again, which might then catch the user out when they move on to perform some other ML maintenance activity.

One possible solution might be that the listener for `prepareForIndicesUpgrade` takes an `Object` instead of a `Boolean`.  Then that `Object` is passed as an extra `preUpgradeResult` argument to `indicesUpgradeComplete`.  This would enable each system indices plugin to remember some information learnt during `prepareForIndicesUpgrade` and make use of it in `indicesUpgradeComplete`.

Maybe there's a better solution but I think it will be necessary to have _some_ way for something to be remembered between the two calls.

Reply:
That's a good point, thanks. There's an extra wrinkle in that we might have a node failure during the process, which would mean we'd need to store that object in a persistent way, which is definitely not impossible and would also allow us to ensure that `prepareForIndicesUpgrade` is only called once per feature (except in the case of a node failure between calling `prepareForIndicesUpgrade` and persisting the object, but I'm not sure that one's avoidable). I'll work on making that change.

Reply:
I think the consequence of having to store it persistently is that it'll be a `Diffable` instead of an `Object`, but close enough. If we want to make things really simple it could just be a `Map<String, String>` or similar.